larpandoracontent/LArReclustering/ThreeDReclusteringAlgorithm.cc

```diff
@@ -0,0 +1,472 @@
+        
+        iCluster++;
+    }
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::SaveList<Pfo>(*this, newPfoListName, m_PfosForReclusteringListName));
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode ThreeDReclusteringAlgorithm::RebuildPfo(const Pfo *pPfoToRebuild, ClusterList &newClustersList)
+{
+    m_newClustersUMap.clear();
+	m_newClustersVMap.clear();
+	m_newClustersWMap.clear();
+
+    //For each of the new 3D clusters, build a new 2D cluster in each view, and a new Pfo
+	PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, this->BuildNewTwoDClusters(pPfoToRebuild, newClustersList));
+	PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, this->BuildNewPfos(pPfoToRebuild, newClustersList));
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool ThreeDReclusteringAlgorithm::PassesCutsForReclustering(const pandora::ParticleFlowObject *const pShowerPfo)
+{
+    if (!LArPfoHelper::IsShower(pShowerPfo)) return false;
+    ClusterList clusterList3D;
+    LArPfoHelper::GetThreeDClusterList(pShowerPfo, clusterList3D);
+
+    if (clusterList3D.empty()) return false;
+    CaloHitList caloHitList3D;
+    clusterList3D.front()->GetOrderedCaloHitList().FillCaloHitList(caloHitList3D);
+
+    //Quality cuts
+    if (caloHitList3D.size() < 2) return false;
+
+    //Some pfos are shower-like and yet include track-like 3D clusters. For the moment I don't want to deal with these.
+    const ClusterList *pShowerClusters(nullptr);
+    PandoraContentApi::GetList(*this, "ShowerClusters3D", pShowerClusters);
+    if(!pShowerClusters) return false;
+
+    if(pShowerClusters->end() == std::find(pShowerClusters->begin(), pShowerClusters->end(), clusterList3D.front())) return false;
+    if(this->GetFigureOfMerit(caloHitList3D)<m_fomThresholdForReclustering) return false; 	
+    return true;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+float ThreeDReclusteringAlgorithm::GetCheatedFigureOfMerit(CaloHitList mergedClusterCaloHitList3D)
+{
+    std::map<int,int> mainMcParticleMap;
+    for (const CaloHit *const pCaloHit : mergedClusterCaloHitList3D)
+    {
+        const CaloHit *const pParentCaloHit = static_cast<const CaloHit *>(pCaloHit->GetParentAddress());
+        int mainMcParticleIndex = this->GetMainMcParticleIndex(pParentCaloHit);
+        std::map<int, int>::iterator it = mainMcParticleMap.find(mainMcParticleIndex);
+ 
+        if (it != mainMcParticleMap.end()) {
+            it->second++;
+        }
+        else {
+            mainMcParticleMap.insert(std::make_pair(mainMcParticleIndex, 1));
+        }
```

Comment:
Don't need braces here? If leaving them, opening braces should be on the following lines?